doc.rkt

```diff
@@ -0,0 +1,274 @@
+#lang racket/base
+
+(require "check-syntax.rkt"
+         "msg-io.rkt"
+         "responses.rkt"
+         "interfaces.rkt"
+         "scheduler.rkt"
+         "editor.rkt"
+         "path-util.rkt"
+         "doc-trace.rkt"
+         racket/match
+         racket/class
+         racket/set
+         racket/list
+         racket/string
+         data/interval-map
+         syntax-color/module-lexer
+         syntax-color/racket-lexer
+         json)
+
+(struct Doc
+  (text
+   trace
+   uri
+   during-batch-change?
+   checked?)
+  #:transparent #:mutable)
+
+(define (send-diagnostics doc diag-lst)
+  (display-message/flush (diagnostics-message (Doc-uri doc) diag-lst)))
+
+;; the only place where really run check-syntax
+(define (doc-run-check-syntax doc)
+  (define (task)
+    (set-Doc-checked?! doc #f)
+    (match-define (list new-trace diags)
+      (check-syntax (uri->path (Doc-uri doc)) (Doc-text doc) (Doc-trace doc)))
+    (send-diagnostics doc diags)
+    (set-Doc-trace! doc new-trace)
+
+    (set-Doc-checked?! doc #t))
```

Comment:
Could this function be called in concurrency? If so, then the `checked?` modification is not safe, need a lock here

Reply:
No. There's only one `task` running for a given document at anytime.
The `set-Doc-*` methods are expected to be safe.